torchgen/gen.py

```diff
@@ -152,6 +156,152 @@ def construct_mapping(self, node, deep=False):  # type: ignore[no-untyped-def]
 ParsedYaml = namedtuple("ParsedYaml", ["native_functions", "backend_indices"])
 
 
+def pre_group_native_functions(
+    native_functions: Sequence[NativeFunction],
+) -> Dict[FunctionSchema, Dict[SchemaKind, NativeFunction]]:
+    pre_grouped_native_functions: Dict[
+        FunctionSchema, Dict[SchemaKind, NativeFunction]
+    ] = defaultdict(dict)
+    for f in native_functions:
+        d = pre_grouped_native_functions[f.func.signature()]
+        assert f.func.kind() not in d
+        d[f.func.kind()] = f
+    return pre_grouped_native_functions
+
+
+# This function is responsible for adding generated NativeFunctions which don't appear
+# explicitly in the codegen.
+# You can inspect the full list of NativeFunctions yourself with the torchgen package, by running
+# torchgen.parse_native_yaml("aten/src/ATen/native/native_functions.yaml", "aten/src/ATen/native/tags.yaml")
+# (Maybe we should make a friendly API for this)
+# Today, the functions that we generate of consist of:
+# - For any *CompositeExplicitAutograd* operators that are missing a functional or out= variant,
+#   we generate those missing variants in terms of the existing ones. This has two benefits:
+#   (1) Allows us to more consistently group together variants into NativeFunctionsGroup objects.
+#   (2) Gives the functionalization pass functional variants to work with, so we don't need to
+#       manually implement functional variants of all operators to get support for all mutable operators.
+def add_generated_native_functions(
+    rs: List[NativeFunction],
+    indices: Dict[DispatchKey, Dict[OperatorName, BackendMetadata]],
```

Comment:
Warning: this input gets mutated! It might be less misleading to mutate both inputs, or none of them.

Reply:
good point - the mutating feels ok to do since it should basically be the first thing that happens after yaml parsing (so anyone using codegen sees the final result), but I'll also mutate `rs` for consistency

Reply:
yes i like mutating